### PR TITLE
Sprite and Animation refactor

### PR DESCRIPTION
**Pull request by juanitogan**
My apologies for such a large PR.  But, well, you can't refactor some things without going big.  Especially, with docs and examples to update too.  It will probably be easier to just look at my fork (especially the doc changes) instead of the diffs.

This PR will break all Flame projects before it.

### Highlights of the changes:

* All Sprite and Animation constructors refactored with focus on simplifying, clarifying, and flexibility.  Most properties are now included as optional named parameters in the constructors.  (Modeled loosely after Qt's [AnimatedSprite](https://doc.qt.io/qt-5/qml-qtquick-animatedsprite.html).)  All async instantiations have been moved to Future<> methods.  (I didn't touch SVG and some other edge stuff much -- someone else can bring things I didn't get to up to speed.)

* The flexibility from the above changes eliminates the need for the animation `sequence` constructor... as well as the entire `spritesheet` module.  Admittedly, the new way can get a bit wordy (as seen in my updated `spritesheet` example app) but can also be beautifully simple (as seen in my updated `particles` example app).

* All image slicing now happens via integer parameters instead of doubles.  This drove me a bit crazy before that I had to do counter-intuitive things like `width: image.height.toDouble()` to slice an image for an animation.  Don't worry, the sprites and animations are still positioned and sized with doubles.

* Because I was breaking so much, I went a little further than normal in renaming a few minor things. `textureWidth` etc. are now `frameWidth` etc.  `currentIndex` is now `currentFrame` (and the previous `currentFrame` getter was removed).  `getSprite()` is now `getCurrentSprite()`.  Others I'm probably forgetting (like what `frameCount` was before).

* `Animation.reversed()` is gone.  See the code notes but, in short, a new and fancy `Animation.reverse` mechanism was built.

* `Animation.paused` feature added.  Particularly useful for those that need to control the `currentFrame` prop in code.

* New `Animation` methods: `advance()`, `pause()`, `resume()`.

* `Flame.images.fromCache(fileName)` method added to encourage the new use pattern.

* `Game.scale` feature added for universal scaling of low-res games.  `tapable.dart` was also changed to make this happen -- to allow for scaling the gesture.  More probably needs to be done with scaling gestures if anyone has some good ideas around it.

* I spent a lot of time rewriting parts of `README.md`, `images.md`, and `components.md` as well.  (And a bit of time fixing the examples I broke.)

I think that covers it, more or less.  I can't say I tested everything I changed but I'm pretty confident 98% of it is solid.  (I didn't see an easy way to test all the examples--they don't just run in Android Studio--so I just relied on the precompiler to check those changes.)

The "why" of all this.  I wrote a small game in Qt recently to see what QtQuick and QML were like.  I then discovered the licensing troubles with publishing Qt stuff so I decided to see what Flutter was like by porting that same game to it.  The Flame port has been more difficult and more time-consuming than writing the original game in Qt.  Thus, I took a long look at what I was doing wrong and where the docs and articles were misleading me.  After refactoring my early Flame test objects about 4 times -- while also rewriting Flame stuff -- this is where I have landed.  I'm sure not everybody will be happy with these changes but, well, I had to try.  I can always keep my own fork under another name if needed.

Beyond all this.  To quote what I wrote in #222 earlier:  While my Flame changes are breaking, they don't change the overall design pattern much.  I haven't gone that far yet even though I would like to see a different composite-game-entity design (the ComposedComponent mixin is good enough for me for now) and maybe an ECS like [this](https://github.com/sschmid/Entitas-CSharp).  What Flame calls a component, I call an entity or game object, and every base entity should be compositive of other entities.  The components would then be things like sprites, animations, and positions/transforms that make up the entity.

Anyhow, thank you for Flame and thank you for reading this.

### CODE REVIEW

File: lib/spritesheet.dart [hunk 1]
```diff
@@ -1,58 +0,0 @@
-import 'sprite.dart';
```
Comment (erickzanardo) [inline lib/spritesheet.dart]: Why remove this class?

This class is a great utility to extract sprites and animations from a single spritesheet image without having to manually calculating the coordinates.

**Comment by erickzanardo** [comment]
btw, rebase your PR to the develop branch.

**Comment by luanpotter** [comment]
I agree with @erickzanardo this is a very big PR and would take me a while to fully review everything. If we could start extracting smaller independent changes to other PRs, would be much nicer to manage, comment and approve/disapprove each suggestion...

**Comment by juanitogan** [comment]
No worries.  I didn't think this PR would get very far.  But, I thought I should at least try my best before spinning off a new game engine.  There's a lot of tl;dr stuff that I didn't bother to write out earlier, explaining all the decisions.  Some of it is in the docs that nobody reads.  The rest, I figured, would come up in the Q&A, if at all.  But, yeah, I can read the writing on the wall.  Probably best to just Blaze my own trail here.  95% of this can't be broken into smaller PRs, so why bother with the other 5%?

Anyhow, before I go, I will at least try to explain why I gave spritesheet.dart the axe (which I knew wouldn't be received well).  Coming into Flutter and Flame as a noob, there were a few terribly confusing things that wasted maybe a week of my time.  One of which was image handling.  It went kinda like this: "Okay, here's how to turn an image into a sprite, I think.  And, oh, here's a different constructor for 1D spritesheets.  And... What the heck?... Why are 2D spritesheets so different than 1D sheets???"  It was all muddy with no clear best practice.  Sure, spritesheet.dart looked like a neat helper class, but was it really necessary?  Was it worth the extra confusion, objects, and memory it leads to?

From there, I set about making `.fromImage` as multipurpose as reasonable for both single-image files and various types of spritesheets (including ones with sprites of different sizes--as long as the frames are still uniform across a particular sprite).  Then, from there, I refocused the nameless constructors on their highest-level child objects.  Those things then cascaded into a pretty wide constructor redux.  After that, with so much now broken, I then decided I might as well take advantage of the breaks to rename a few other things.

All that left spritesheet.dart redundant and hanging in limbo.  I considered rewriting it but couldn't justify the confusion it adds versus the benefits.  For example, (to oversimplify with pseudo-code a bit):

* `getSprite(2, 3)` becomes:
* `.fromImage(img, frameX: frameWidth * (2-1), frameY: frameHeight * (3-1))`

And

* `createAnimation(2, from: 3, to: 5)` becomes:
* `.fromImage(img, frameX: frameWidth * 1, frameY: frameHeight * 2, frameCount: 3)` or
* `.fromImage(img, firstFrame: 21, frameCount: 3)` (assuming a 10x10 sheet), or maybe
* `.fromImage(img, frameY: frameHeight * 2, firstFrame: 1, frameCount: 3)`

Whatever works for your load loops.  So, yeah, wordier at times with math requiring a couple more brain cells, but less wordy at other times and in the long run.  Personally, I feel that anyone who gives this a fair try for few days won't feel a need for any helper classes.  And, if they really want one, they can make one pretty easily that works however they want.

The other confusing things I ran into as a noob (that sucked up most of my time) I was intending to address in the docs more, but I haven't gotten very far on that yet.  I'll probably now just take the plunge to refactor even heavier and make a new spinoff game engine to support my games.  Perhaps call it Pogo after the game I'm porting.  We can then cherry pick what we like from each other's projects.  Although, I don't have much time for any more open-source work so I doubt I'll get far (I need to spend most of my time looking for paid work right now).

There's a lot great work here.  No doubt.  But, I also see the growing pains.  I'm trying to make it easier for noobs to ramp up on, and trying to make it more of a pleasure for odd programmers like me to work with.  More concise, more self-apparent, better lingo, etc.  I loved how easy it was to grok Qt for simple game making.  Tried to do a similar quick grok with Flutter and Flame and just didn't succeed.  Sorry.

**Comment by kurtome** [comment]
I'm not a maintainer on this project, but just to add my two cents...

A lot of these API changes seem good, but making this many breaking changes without a migration plan is a recipe for a PR that's never going to get merged. Instead it would make a lot more sense to add newer better APIs in a backwards compatible way. For example, you could add a lot of these cleaned up constructors while leaving the old constructors/classes but marking them as deprecated and adding comments recommending the new API.

Obviously you can fork the project and do your own thing, but if you want to help other people use better tools (which it sounds like you do) you'll have a lot more impact working with the project that's already popular (Flame in this case).

**Comment by juanitogan** [comment]
@kurtome  Thank you for the comments.  I agree with all of what you said.  There is some evidence left in the code that, early on, I tried looking for a migration path.  But, eventually, I ran into too much that was getting too weird, so I gave up on that and then just shifted things as I needed.  (All the API changes to sprite.dart, animation.dart, component.dart, and animation_component.dart are not outlined above -- you need to go to the code and docs for that, where, hopefully, the complexity of the shift becomes more apparent.)  I could have gone further with some petty things (like done() vs finished(), for example) but didn't bother in an attempt to avoid trampling on things that are simply opinion based.

Personally, I don't want to rock the boat too much.  If everybody likes Flame the way it is, that's great.  Keep it as is.  I respect the work done here and will give credit to it.  But, I'm not all sunshine and flowers either -- I believe in the necessity of constructive criticism -- even if it is a tectonic shift.  I won't be offended if this PR isn't approved (and it seems it won't).  This was just me taking a few days to make a small attempt to give something back.  I'm cool with it wherever it lands.

So while, yes, I do like to make tools others find useful... really, ultimately, I am primarily focused on publishing a game and not a game engine.  I have bills to pay and/or a resume to pad before I can focus on my open-source stuff.  Second to that, I have a mantra that says something like: Dare greatly... but don't be a jerk about it either.

So, in other words, I don't hesitate much to build or change tools to how I need them.  Sometimes I take time to share those tools.  Sometimes I don't.  (For example, I built a pretty awesome Bezier placement tool for my Unity game, GRITS Racing, that does things I couldn't find in other tools.  I might like to share it but I doubt I'll ever find time to polish enough to do so.  On the other hand, I just might adapt it to a new Bezier alignment or envelope tool for InkScape someday since I care more about that for some reason.)

I expect I'll take some middle route with Flame.  I've taken the hesitation here as a signal to forget stepping carefully and just lean even more into changing Flame to how I will find it most useful.  Thus, I'm now looking harder at changing the component system like I've hinted before.  I don't think I have time to build it into a bona fide ECS, but I can push it into more of a pseudo ECS fairly easily (which is easier than an ECS for noobs to ramp up on anyhow).  I'm now free to do radical things like renaming Animation to AnimationComponent, and AnimationComponent to AnimationGameEntity, or whatever makes most sense (AnimationEntity might just be a helper to GameEntity).  This is where I burned up most of my time -- trying to understand Flame's design pattern(s) (I see examples and guides that go too many different ways) -- it was not with the Sprite and Animation constructors I rewrote.  I'm also looking suspiciously at Position vs the Vector2 in Box2D (from vector_math).

From there, I will probably make everything I've changed available under a new name with credit to Flame.  But, at the same time, I know this game engine is already too big for me to maintain alone.  Thus, maybe some will find my semi-dead code useful and help, maybe they won't.  This is not a huge deal to me as, like I said, this is not my focus.  I just need good tools to make good apps, and, as long as those tools work, popularity doesn't matter a whole lot when one-offing a game.  Ideally, someone with more time than I will take my version of the engine and make something even more useful out of it.

Another tl;dr, I know.  That's who I am.  Just trying to be transparent.

**Comment by luanpotter** [comment]
As I said, let's extract smaller pieces and merge if they make sense so we can analyze them better. @erickzanardo @renancaraujo and I will analyze and scrape the bits we like but also feel free to open smaller PRs as we sort this through. So I'm closing this for now as we won't merge this as is, but please feel free to open more directed issues and PRs and we will be glad to continue the conversation.

**Comment by juanitogan** [comment]
No worries.  I refactored some of these changes even further and then kept on going into what turned into a total refactor.  After a few weeks of rewriting the docs too, the Pogo game engine has finally been published.  Of course, my recommendation would be to pull everything I just changed back into Flame, but I doubt that'll happen.  Thus: [Pogo](https://github.com/juanitogan/pogo)... in case this design pattern makes sense to anybody besides me.
